File: core/LongestCommonSubsequence/LongestCommonSubsequence.java
package com.dbc;

import java.util.Arrays;

public class LongestCommonSubsequence {
    public int longestCommonSubsequence(String text1, String text2) {
        int[][] dp = new int[text1.length() + 1][text2.length() + 1];
        for (int i = 0; i < text1.length(); i++) {
            Arrays.fill(dp[i], 0);
        }
        for (int row = 1; row < text1.length() + 1; row++) {
            for (int col = 1; col < text2.length() + 1; col++) {
                if (text1.charAt(row - 1) == text2.charAt(col - 1)) {
                    dp[row][col] = dp[row - 1][col - 1] + 1;
                } else {
                    dp[row][col] = Math.max(Math.max(dp[row - 1][col], dp[row][col - 1]), dp[row - 1][col - 1]);
                }
            }
        }
        return dp[text1.length()][text2.length()];
    }
}

File: core/LongestCommonSubsequence/LongestCommonSubsequence.py
def longestCommonSubsequence(text1: str, text2: str) -> int:
    length1, length2 = len(text1), len(text2)
    dp = [[0] * (length2 + 1) for _ in range(length1 + 1)]
    for row in range(1, length1 + 1):
        for col in range(1, length2 + 1):
            if text1[row - 1] == text2[col - 1]:
                dp[row][col] = dp[row - 1][col - 1] + 1
            else:
                dp[row][col] = max(max(dp[row - 1][col - 1], dp[row][col - 1]), dp[row - 1][col])

    return dp[-1][-1]


if __name__ == '__main__':
    print(longestCommonSubsequence(text1 = "abcde", text2 = "ace"))



